Working with Pandas
=================

<div class="overview-this-is-a-title overview">
<p class="overview-title">Overview</p>
<p>Questions</p>
    <ul>
        <li>How can I use import data for analysis in my notebook?
    </ul>
<p>Objectives:</p>
    <ul>
        <li>Import the pandas library.
        <li>Use pandas library funtions to import data from a csv file.
        <li>Import data from a .csv formatted file.
    </ul>
<p>Keypoints:</p>
    <ul>
        <li>Use the pandas library to create dataframes from csv formatted data.</li>
    </ul>
</div>

## What is pandas and why do we use it?

Pandas is a python library that is designed to work with two dimensional data arrays. It is built on numpy, another python library that specializes in numerical analysis. Numpy also has the ability to create and analyze data arrays. If you are not familiar with arrays, here are some simple examples.

### 1D arrays
A one dimensional array is simply a list of items, for example, a list of the elements: H, He, Li, Be, B, etc.

### 2D arrays
A two dimensional array is an array which has rows and columns. It can have any number of columns and rows. A spreadsheet with rows and columns is analogous to a 2D array.

### 3D arrays
A three-dimensional array would be a collection of two dimensional arrays. For example, this might be a collection of x, y, and z coordinates for a structure as a function of time. 

The numpy library has functions that will manage n-dimensional arrays, while pandas only works on 2D arrays. You may need numpy for nD arrays at some point, but for this workshop, we will learn to use pandas as we learn how to perform linear and nonlinear regression of laboratory data based on data in 2D arrays.

The pandas library contains powerful tools for working with 2D data arrays, including the ability to identify the rows and columns of data by unique identifiers: things like "Protein Concentration (mg/mL)" and "initial velocity" rather than "row 1" or "column C"). Pandas also has many more functions that we will not explore in this workshop, but here are three excellent free online resources for learning more about pandas:

+ [Using pandas for data analysis](https://education.molssi.org/python-data-analysis/02-pandas/index.html) from MolSSI
+ Charlie Weiss's excellent online textbook, *Scientific Computing for Chemists*, which you can find on his GitHub site, [SciCompforChemists](https://github.com/weisscharlesj/SciCompforChemists)
+ Corey Schafer's [Pandas Tutorials on YouTube](https://www.youtube.com/watch?v=ZyhVh-qRZPA&list=PL-osiE80TeTsWmV9i9c58mdDCSskIFdDS)

## Importing python libraries

As we move toward data analysis, we will need to import a series python **libraries**. A **library** is a set of modules which contain related functions that can be used to complete specific tasks. Using libraries in Python reduces the amount of code you have to write. Usually a function has some type of input and gives a particular output.  To use a function that is in a library, you often use the dot notation introduced previously.

In the last lesson, we imported the `os` library, which can be used to assign the location of a file to a variable (e.g. datafile), so our script can be transferred between computers with different operating systems (e.g., Mac vs Windows). In this lesson, we will be using the `numpy` and `pandas`, and we will add a few more libraries in the future. As we proceed, we'll build this table. 

| Library | Uses | Abbreviation |
| :------- | :----: | :------------: |
| os | file management in operating systems | os |
| numpy | calculations | np  | 
| pandas | data management | pd |

To start we will import the two libraries we need to complete the work in this notebook, os and pandas.

In [1]:
import os
import pandas as pd

### Locating the data

To work with data in Jupyter notebooks, you must first find the data. In the file parsing lesson, we used the `pwd` (print working directory) command to find out where we were located on our computers. In this lesson, we will learn another command, `ls`, which gives us a listing of the files and folders in the working directory.

In [2]:
ls

Bradford_plot.png
Bradford_plot2.png
Bradford_plot3.png
Creating_Plots_in_Jupyter_Notebooks.ipynb
EC_class.txt
EC_class2.txt
File_Parsing.ipynb
Linear_Regression.ipynb
Processing_Multiple_Files_and_Writing_Files.ipynb
Untitled.ipynb
Working_with_Pandas.ipynb
data/
data.zip
images/
introduction.ipynb
nonlinear_regression_part_1.ipynb
nonlinear_regression_part_2.ipynb
resolutions.txt
setup.ipynb
workshop_schedule.ipynb


Notice that `data/` appears in a different font and is followed by a forward slash. That is because `data` is a folder or directory. In fact, the data folder contains the file we want to use. To get a listing of the files in the `data` folder, you can just type `ls data`. The `ls` command is unix/linux is similar to the `dir` command when you are working at the command prompt in Windows.

In [3]:
ls data

AP_kin.csv                  chymotrypsin_kinetics.csv
AP_kinetics.csv             chymotrypsin_kinetics.xlsx
Ground_water.csv            enzyme_kinetics.csv
MM_data.csv                 enzyme_kinetics.xlsx
MM_data1.csv                protein_assay.csv
MM_data_for_NLRpt2.csv      protein_assay2.csv
PDB678.csv                  protein_samples.csv
PDB_files/                  thrombin_with_ligands.csv
chymo_MM_data.csv


You will notice a number of files and folders in the data folder. We will be working with the file, `thrombin_with_ligands.csv`. As a first step, we will assign the location of that file to a variable, using the path.join function from the `os` library. 

In [4]:
thrombin_file = os.path.join('data', 'thrombin_with_ligands.csv')
print(thrombin_file)

data/thrombin_with_ligands.csv


### Importing the data with pandas
As mentioned above, the pandas library has a number of excellent tools for importing data. We will be using the `read_csv` function from the pandas library to import our data. You may have encountered csv files before; csv stands for comma separated values. This is a format that spreadsheet programs like Excel and Google Sheets can both read and write. If you open a csv file in a text editor, it simply contains rows of information, where the information is separated by commas. The information can be integers, real numbers or strings. Here are the first four rows of the csv file that we will be exploring using pandas.

```
PDB ID,Method,Resolution,Structure,Ligand ID,Ligand name

3SHC,X-RAY DIFFRACTION,1.9,Human Thrombin,B01,D-phenylalanyl-N-[(4-chloropyridin-2-yl)methyl]-L-prolinamide

3P17,X-RAY DIFFRACTION,1.43,Thrombin,99P,D-phenylalanyl-N-(pyridin-3-ylmethyl)-L-prolinamide

2ZNK,X-RAY DIFFRACTION,1.8,Thrombin,31U,D-leucyl-N-(4-carbamimidoylbenzyl)-L-prolinamide
```

If you look at these data carefully, you can see how they could easily be arranged in a table, with the labels in the first row as the column headers. Before we can explore how pandas can do that for us, let's look at the help command for the `read_csv` function in pandas.

```python
help(pd.read_csv)
```

Notice the syntax for the command:  help is followed by parentheses, which contain the function name that we want to explore. Also note that `pandas` is abbreviated as `pd`. That may seem strange at first, but you will get used to library abbreviations in short order if you work much with Jupyter notebooks, or with python in general.

If you enter this command in a cell, the output will be over 200 lines long. You can learn about all of the arguments that can be attached to the read_csv function and how they affect the dataframe. Just to give you a taste, here are the first few lines you would see.

Help on function read_csv in module pandas.io.parsers:

```python
read_csv(filepath_or_buffer: Union[str, pathlib.Path, IO[~AnyStr]], sep=',', delimiter=None,
header='infer', names=None, index_col=None, usecols=None, squeeze=False, prefix=None, mangle_dupe_cols=True, dtype=None, engine=None, converters=None, true_values=None, false_values=None, skipinitialspace=False, skiprows=None, skipfooter=0, nrows=None, na_values=None, keep_default_na=True, na_filter=True, verbose=False, skip_blank_lines=True, parse_dates=False, infer_datetime_format=False, keep_date_col=False, date_parser=None, dayfirst=False, cache_dates=True, iterator=False, chunksize=None, compression='infer', thousands=None, decimal: str = '.', lineterminator=None, quotechar='"', quoting=0, doublequote=True, escapechar=None, comment=None, encoding=None, dialect=None, error_bad_lines=True, warn_bad_lines=True, delim_whitespace=False, low_memory=True, memory_map=False, float_precision=None)
```

The panda library contains many more options than we have time to explore right now, but one thing to pay special attention to is called a `dataframe`. The [pandas dataframe](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) is the primary data structure that is produced and manipulated by pandas functions. We can take an existing 2D data array and tell pandas to convert it to a DataFrame, or we can use the read_csv command to convert a csv file to a pandas dataframe. As our example, we will be exploring a csv file containing information about different thrombin structures found in the [Protein Data Bank](https://www.rcsb.org/). So let's create our DataFrame and explore what it contains.

In [5]:
thrombin_df = pd.read_csv(thrombin_file) #thrombin_df is a convenient name for our dataframe.
thrombin_df

PDB ID             Method  Resolution        Structure Ligand ID  \
0    3SHC  X-RAY DIFFRACTION        1.90   Human Thrombin       B01   
1    3P17  X-RAY DIFFRACTION        1.43         Thrombin       99P   
2    2ZNK  X-RAY DIFFRACTION        1.80         Thrombin       31U   
3    3SI3  X-RAY DIFFRACTION        1.55   Human Thrombin       B03   
4    3SI4  X-RAY DIFFRACTION        1.27   Human Thrombin       B04   
5    3SV2  X-RAY DIFFRACTION        1.30   Human Thrombin       P05   
6    1UVU  X-RAY DIFFRACTION        2.80  Bovine Thrombin       DCH   
7    1UVT  X-RAY DIFFRACTION        2.50  Bovine Thrombin       I48   
8    2C8Z  X-RAY DIFFRACTION        2.14         Thrombin       C2A   
9    2C8Y  X-RAY DIFFRACTION        2.20         Thrombin       C3M   
10   2C90  X-RAY DIFFRACTION        2.25         Thrombin       C1M   

                                          Ligand name  
0   D-phenylalanyl-N-[(4-chloropyridin-2-yl)methyl...  
1   D-phenylalanyl-N-(pyridin-3-ylmethyl)-L-prolin...  
2    D-leucyl-N-(4-carbamimidoylbenzyl)-L-prolinamide  
3   D-phenylalanyl-N-(pyridin-2-ylmethyl)-L-prolin...  
4   D-phenylalanyl-N-[(1-methylpyridinium-2-yl)met...  
5   D-phenylalanyl-N-(pyridin-4-ylmethyl)-L-prolin...  
6   3-(7-DIAMINOMETHYL-NAPHTHALEN-2-YL)-PROPIONIC ...  
7   N-{3-METHYL-5-[2-(PYRIDIN-4-YLAMINO)-ETHOXY]-P...  
8                       1-(3-CHLOROPHENYL)METHANAMINE  
9   N-[(2R,3S)-3-AMINO-2-HYDROXY-4-PHENYLBUTYL]NAP...  
10                    1-(4-CHLOROPHENYL)-1H-TETRAZOLE

### Finding information in a dataframe

The dataframe appears as a table. The data are in rows ("indexes") and columns ("series"). If you want to look at the data in a single column (one series), you can use this command.

In [6]:
thrombin_df['Resolution']

0     1.90
1     1.43
2     1.80
3     1.55
4     1.27
5     1.30
6     2.80
7     2.50
8     2.14
9     2.20
10    2.25
Name: Resolution, dtype: float64

<div class="exercise-this-is-a-title exercise">
<p class="exercise-title">Check your understanding</p>
    <p>How can you print out the ligand names for each of the structures? </p>
    
```{admonition} Solution
:class: dropdown

```python
thrombin_df['Ligand name']
```
    
</div>

You can use the loc function from the pandas library to get all the information from a single index or row.

In [7]:
thrombin_df.loc[2]

PDB ID                                                     2ZNK
Method                                        X-RAY DIFFRACTION
Resolution                                                  1.8
Structure                                              Thrombin
Ligand ID                                                   31U
Ligand name    D-leucyl-N-(4-carbamimidoylbenzyl)-L-prolinamide
Name: 2, dtype: object

You can also use the loc command to locate the information from a specific cell. If you enter the index (row) of 2 and the column 'Resolution', the resolution for the structure in the third row is returned. You may be thinking, "Resolution is the third column. Why does python think it is column 2?" Remember that python starts counting at 0. So the number for the first column is 0 and the number for the third column is 2.

In [8]:
thrombin_df.loc[2, 'Resolution']

1.8

There are times when you may want to use the index number and column number to locate the information in that cell in the dataframe. In this case, you can use the iloc (for integer locate) command, which accepts integers in the form of \[row, column]. 

In [9]:
thrombin_df.iloc[2,2]

1.8

If you only want to look at the first five rows of a dataframe, you can use the `head` function; you can use the `tail` function to look at the last five rows. We always include () after the name of the function. This is how we can pass arguments into the function. Try using the argument 3 in the head command.

In [10]:
thrombin_df.head()

PDB ID             Method  Resolution       Structure Ligand ID  \
0   3SHC  X-RAY DIFFRACTION        1.90  Human Thrombin       B01   
1   3P17  X-RAY DIFFRACTION        1.43        Thrombin       99P   
2   2ZNK  X-RAY DIFFRACTION        1.80        Thrombin       31U   
3   3SI3  X-RAY DIFFRACTION        1.55  Human Thrombin       B03   
4   3SI4  X-RAY DIFFRACTION        1.27  Human Thrombin       B04   

                                         Ligand name  
0  D-phenylalanyl-N-[(4-chloropyridin-2-yl)methyl...  
1  D-phenylalanyl-N-(pyridin-3-ylmethyl)-L-prolin...  
2   D-leucyl-N-(4-carbamimidoylbenzyl)-L-prolinamide  
3  D-phenylalanyl-N-(pyridin-2-ylmethyl)-L-prolin...  
4  D-phenylalanyl-N-[(1-methylpyridinium-2-yl)met...

In [11]:
thrombin_df.head(3)

PDB ID             Method  Resolution       Structure Ligand ID  \
0   3SHC  X-RAY DIFFRACTION        1.90  Human Thrombin       B01   
1   3P17  X-RAY DIFFRACTION        1.43        Thrombin       99P   
2   2ZNK  X-RAY DIFFRACTION        1.80        Thrombin       31U   

                                         Ligand name  
0  D-phenylalanyl-N-[(4-chloropyridin-2-yl)methyl...  
1  D-phenylalanyl-N-(pyridin-3-ylmethyl)-L-prolin...  
2   D-leucyl-N-(4-carbamimidoylbenzyl)-L-prolinamide

In [12]:
thrombin_df.tail()

PDB ID             Method  Resolution        Structure Ligand ID  \
6    1UVU  X-RAY DIFFRACTION        2.80  Bovine Thrombin       DCH   
7    1UVT  X-RAY DIFFRACTION        2.50  Bovine Thrombin       I48   
8    2C8Z  X-RAY DIFFRACTION        2.14         Thrombin       C2A   
9    2C8Y  X-RAY DIFFRACTION        2.20         Thrombin       C3M   
10   2C90  X-RAY DIFFRACTION        2.25         Thrombin       C1M   

                                          Ligand name  
6   3-(7-DIAMINOMETHYL-NAPHTHALEN-2-YL)-PROPIONIC ...  
7   N-{3-METHYL-5-[2-(PYRIDIN-4-YLAMINO)-ETHOXY]-P...  
8                       1-(3-CHLOROPHENYL)METHANAMINE  
9   N-[(2R,3S)-3-AMINO-2-HYDROXY-4-PHENYLBUTYL]NAP...  
10                    1-(4-CHLOROPHENYL)-1H-TETRAZOLE

### Sorting data with pandas
You can also sort the data using the pandas sort_values function. To sort the thrombin_df dataframe into ascending Resolution values, use the following command.

In [13]:
thrombin_df.sort_values('Resolution')

PDB ID             Method  Resolution        Structure Ligand ID  \
4    3SI4  X-RAY DIFFRACTION        1.27   Human Thrombin       B04   
5    3SV2  X-RAY DIFFRACTION        1.30   Human Thrombin       P05   
1    3P17  X-RAY DIFFRACTION        1.43         Thrombin       99P   
3    3SI3  X-RAY DIFFRACTION        1.55   Human Thrombin       B03   
2    2ZNK  X-RAY DIFFRACTION        1.80         Thrombin       31U   
0    3SHC  X-RAY DIFFRACTION        1.90   Human Thrombin       B01   
8    2C8Z  X-RAY DIFFRACTION        2.14         Thrombin       C2A   
9    2C8Y  X-RAY DIFFRACTION        2.20         Thrombin       C3M   
10   2C90  X-RAY DIFFRACTION        2.25         Thrombin       C1M   
7    1UVT  X-RAY DIFFRACTION        2.50  Bovine Thrombin       I48   
6    1UVU  X-RAY DIFFRACTION        2.80  Bovine Thrombin       DCH   

                                          Ligand name  
4   D-phenylalanyl-N-[(1-methylpyridinium-2-yl)met...  
5   D-phenylalanyl-N-(pyridin-4-ylmethyl)-L-prolin...  
1   D-phenylalanyl-N-(pyridin-3-ylmethyl)-L-prolin...  
3   D-phenylalanyl-N-(pyridin-2-ylmethyl)-L-prolin...  
2    D-leucyl-N-(4-carbamimidoylbenzyl)-L-prolinamide  
0   D-phenylalanyl-N-[(4-chloropyridin-2-yl)methyl...  
8                       1-(3-CHLOROPHENYL)METHANAMINE  
9   N-[(2R,3S)-3-AMINO-2-HYDROXY-4-PHENYLBUTYL]NAP...  
10                    1-(4-CHLOROPHENYL)-1H-TETRAZOLE  
7   N-{3-METHYL-5-[2-(PYRIDIN-4-YLAMINO)-ETHOXY]-P...  
6   3-(7-DIAMINOMETHYL-NAPHTHALEN-2-YL)-PROPIONIC ...

Notice that the numbers in the first column (the index values) are not in order. If you type “thrombin_df” now, you will get back the unsorted dataframe. 

In [14]:
thrombin_df

PDB ID             Method  Resolution        Structure Ligand ID  \
0    3SHC  X-RAY DIFFRACTION        1.90   Human Thrombin       B01   
1    3P17  X-RAY DIFFRACTION        1.43         Thrombin       99P   
2    2ZNK  X-RAY DIFFRACTION        1.80         Thrombin       31U   
3    3SI3  X-RAY DIFFRACTION        1.55   Human Thrombin       B03   
4    3SI4  X-RAY DIFFRACTION        1.27   Human Thrombin       B04   
5    3SV2  X-RAY DIFFRACTION        1.30   Human Thrombin       P05   
6    1UVU  X-RAY DIFFRACTION        2.80  Bovine Thrombin       DCH   
7    1UVT  X-RAY DIFFRACTION        2.50  Bovine Thrombin       I48   
8    2C8Z  X-RAY DIFFRACTION        2.14         Thrombin       C2A   
9    2C8Y  X-RAY DIFFRACTION        2.20         Thrombin       C3M   
10   2C90  X-RAY DIFFRACTION        2.25         Thrombin       C1M   

                                          Ligand name  
0   D-phenylalanyl-N-[(4-chloropyridin-2-yl)methyl...  
1   D-phenylalanyl-N-(pyridin-3-ylmethyl)-L-prolin...  
2    D-leucyl-N-(4-carbamimidoylbenzyl)-L-prolinamide  
3   D-phenylalanyl-N-(pyridin-2-ylmethyl)-L-prolin...  
4   D-phenylalanyl-N-[(1-methylpyridinium-2-yl)met...  
5   D-phenylalanyl-N-(pyridin-4-ylmethyl)-L-prolin...  
6   3-(7-DIAMINOMETHYL-NAPHTHALEN-2-YL)-PROPIONIC ...  
7   N-{3-METHYL-5-[2-(PYRIDIN-4-YLAMINO)-ETHOXY]-P...  
8                       1-(3-CHLOROPHENYL)METHANAMINE  
9   N-[(2R,3S)-3-AMINO-2-HYDROXY-4-PHENYLBUTYL]NAP...  
10                    1-(4-CHLOROPHENYL)-1H-TETRAZOLE

However, if you want to keep retain the sorting, you can add `inplace = True` to the command.

In [15]:
thrombin_df.sort_values('Resolution', inplace = True)
thrombin_df

PDB ID             Method  Resolution        Structure Ligand ID  \
4    3SI4  X-RAY DIFFRACTION        1.27   Human Thrombin       B04   
5    3SV2  X-RAY DIFFRACTION        1.30   Human Thrombin       P05   
1    3P17  X-RAY DIFFRACTION        1.43         Thrombin       99P   
3    3SI3  X-RAY DIFFRACTION        1.55   Human Thrombin       B03   
2    2ZNK  X-RAY DIFFRACTION        1.80         Thrombin       31U   
0    3SHC  X-RAY DIFFRACTION        1.90   Human Thrombin       B01   
8    2C8Z  X-RAY DIFFRACTION        2.14         Thrombin       C2A   
9    2C8Y  X-RAY DIFFRACTION        2.20         Thrombin       C3M   
10   2C90  X-RAY DIFFRACTION        2.25         Thrombin       C1M   
7    1UVT  X-RAY DIFFRACTION        2.50  Bovine Thrombin       I48   
6    1UVU  X-RAY DIFFRACTION        2.80  Bovine Thrombin       DCH   

                                          Ligand name  
4   D-phenylalanyl-N-[(1-methylpyridinium-2-yl)met...  
5   D-phenylalanyl-N-(pyridin-4-ylmethyl)-L-prolin...  
1   D-phenylalanyl-N-(pyridin-3-ylmethyl)-L-prolin...  
3   D-phenylalanyl-N-(pyridin-2-ylmethyl)-L-prolin...  
2    D-leucyl-N-(4-carbamimidoylbenzyl)-L-prolinamide  
0   D-phenylalanyl-N-[(4-chloropyridin-2-yl)methyl...  
8                       1-(3-CHLOROPHENYL)METHANAMINE  
9   N-[(2R,3S)-3-AMINO-2-HYDROXY-4-PHENYLBUTYL]NAP...  
10                    1-(4-CHLOROPHENYL)-1H-TETRAZOLE  
7   N-{3-METHYL-5-[2-(PYRIDIN-4-YLAMINO)-ETHOXY]-P...  
6   3-(7-DIAMINOMETHYL-NAPHTHALEN-2-YL)-PROPIONIC ...

## Exploring data with pandas

Now we are going to look at a larger data set (678 serine protease structures from the PDB) to learn about some additional tools for exploring data with pandas. Here are the steps we will follow:
* Import the data for the 678 serine proteases into a new dataframe
* Check to make sure that the numbers (temperature and resolution) are floats, not strings
* Convert the numbers to floats if need be
* Find the mean and median values for resolution
* Find the most common value for temperature
* Find all structures with a resolution between 1.4 and 1.8 that were determined a temperatures at or below 125K

In [16]:
# Import the data using the os library

protease_file = os.path.join('data', 'PDB678.csv')
protease_df = pd.read_csv(protease_file)
protease_df.head()

Entry ID  Collection Temperature Release Date  Resolution  \
0     2R69                   100.0     12/25/07         3.8   
1     2QTW                   100.0      9/18/07         1.9   
2     2R9P                   100.0     12/11/07         1.4   
3     2QY0                   100.0       2/5/08         2.6   
4     2QXJ                   100.0       1/8/08         2.1   

                                               Title  
0  Crystal structure of Fab 1A1D-2 complexed with...  
1  The Crystal Structure of PCSK9 at 1.9 Angstrom...  
2  Human mesotrypsin complexed with bovine pancre...  
3  Active dimeric structure of the catalytic doma...  
4  Crystal Structure of Human Kallikrein 7 in Com...

### Checking the data type

Previously when we checked for a data type for a value that is assigned to a variable, we used this type of command:

```python
type(variable)
```

Let's try this with one of the series from our new dataframe.

In [17]:
type(protease_df['Collection Temperature'])

pandas.core.series.Series

This tells us that the Collection Temperature is a series in the protease_df dataframe, but we need to know the data type for the values in that series. To do this, we have a new command for working with series in dataframes:

```python
df['Series Header'].dtype
```

In [18]:
# Check to make sure that the temperature values are floats, not strings

protease_df['Collection Temperature'].dtype

dtype('float64')

In [19]:
# Check to make sure that the resolution values are floats, not strings

protease_df['Resolution'].dtype

dtype('float64')

So the numbers in those series in the dataframe are all floats, so we there is no need to convert them. 

A couple more points on data type in pandas dataframes. 

* We can actually check the data types for all of the series in the dataframe with a single command:   

```python
df.dtypes
```

Notice that we use the plural `dtypes` to get the types for all of series in the dataframe.

* You can use this command when it is necessary to convert the data type for values in a dataframe series from string to float:  

```python
df['string_col'] = df['string_col'].astype('float')
```

In [20]:
protease_df.dtypes

Entry ID                   object
Collection Temperature    float64
Release Date               object
Resolution                float64
Title                      object
dtype: object

Now we will use a few built-in pandas functions to explore the data for resolution and  temperature. Notice that the function is followed by open and closed parentheses. In these examples, no arguments are passed to the function, but they could be passed inside the parentheses.

In [21]:
# Determine the median value for resolution

protease_df['Resolution'].median()

1.93

In [22]:
# Determine the mean (mathematical average) for resolution

protease_df['Resolution'].mean()

2.0005582554697194

In [23]:
# Determine the standard deviation for resolution

protease_df['Resolution'].std()

0.607453279532895

In [24]:
protease_df['Collection Temperature'].mode()

0    100.0
dtype: float64

Next we are going to query the dataframe looking for specific items. We can use comparison operators (>, <, ==, !=) to select the rows that we want from the dataframe.

In [25]:
# Query for structure with resolution > 1.4 Angstroms

protease_df.query('Resolution > 1.4')

Entry ID  Collection Temperature Release Date  Resolution  \
0       2R69                   100.0     12/25/07        3.80   
1       2QTW                   100.0      9/18/07        1.90   
3       2QY0                   100.0       2/5/08        2.60   
4       2QXJ                   100.0       1/8/08        2.10   
5       2R9J                   300.0      10/9/07        2.55   
..       ...                     ...          ...         ...   
671     7JQO                   100.0      7/21/21        1.60   
673     7JRX                   100.0      7/21/21        1.77   
674     7JTQ                   100.0      12/9/20        3.50   
675     7E6J                   100.0      6/30/21        1.90   
676     7DZD                   100.0      7/14/21        2.00   

                                                 Title  
0    Crystal structure of Fab 1A1D-2 complexed with...  
1    The Crystal Structure of PCSK9 at 1.9 Angstrom...  
3    Active dimeric structure of the catalytic doma...  
4    Crystal Structure of Human Kallikrein 7 in Com...  
5    Ligand recognition in C-lobe: The crystal stru...  
..                                                 ...  
671  Crystal structure of the R64D mutant of Bauhin...  
673  Crystal structure of the R64F mutant of Bauhin...  
674  Human Complement Factor B Inhibited by a Slow ...  
675  Aspartyl/Asparaginyl beta-hydroxylase (AspH) H...  
676  Crystal structure of uPA in complex with cleav...  

[582 rows x 5 columns]

That query gave lots of results: 582 of the 678 structures have resolution values greater than 1.4 Angstroms. Let's make the query more selective - for structures with resolutions in a specific range. In the following cell, we'll try to find structures that were determined at a temperature of 100K.

In [26]:
# Query for structure with resolution between 1.4 and 1.8 Angstroms

protease_df.query('Resolution > 1.4' and 'Resolution < 1.8')

Entry ID  Collection Temperature Release Date  Resolution  \
2       2R9P                   100.0     12/11/07       1.400   
6       2QXI                   100.0       1/8/08       1.000   
10      2RA3                   100.0     12/11/07       1.460   
14      2SGF                   287.0      8/26/03       1.750   
25      6DCM                   100.0      5/15/19       1.031   
..       ...                     ...          ...         ...   
664     7CRA                    80.0      5/26/21       1.700   
670     7JQN                   100.0      7/21/21       1.500   
671     7JQO                   100.0      7/21/21       1.600   
672     7JQK                   100.0      7/21/21       1.330   
673     7JRX                   100.0      7/21/21       1.770   

                                                 Title  
2    Human mesotrypsin complexed with bovine pancre...  
6    High resolution structure of Human Kallikrein ...  
10   Human cationic trypsin complexed with bovine p...  
14   PHE 18 VARIANT OF TURKEY OVOMUCOID INHIBITOR T...  
25   The atomic resolution crystal structure of Kri...  
..                                                 ...  
664  Crystal structure of the N-terminal fragment (...  
670  Crystal structure of the R64M mutant of Bauhin...  
671  Crystal structure of the R64D mutant of Bauhin...  
672  Crystal structure of the R64A mutant of Bauhin...  
673  Crystal structure of the R64F mutant of Bauhin...  

[253 rows x 5 columns]

In [27]:
# Query for structures where the data were collected at 100K

protease_df.query('Collection Temperature == 100')

SyntaxError: invalid syntax (<unknown>, line 1)

So now we have an error. If you read the last few lines of the error report, you see that the query structure does not accept a series label (Collection Temperature) because it contains a space. When I encountered this, I just did an Internet search with the terms "query pandas database columns with labels containing spaces" and one of the top hits from [Stack Overflow](
https://stackoverflow.com/questions/50697536/pandas-query-function-not-working-with-spaces-in-column-names) explained that you can put back-ticks around the series label. Back-ticks can be found on the upper left hand most key on some keyboards. They look like a backwards leaning apostrophe - \`\. When we surrond the series label, Collection Temperature with back-ticks, we get the desired results. Please note that the entire expression within the parentheses is also enclosed in single quotes.

In [ ]:
# Second attempt to query for structures where the data were collected at 100K

protease_df.query('`Collection Temperature` == 100')

Suppose you wanted to save this subset of the protease_df dataframe to a new dataframe. Here is the syntax to do that.

```python
new_df = df.query('series label')
```

We'll execute this command in the next cell to create a new dataframe that contains only structures whose data were collected at 100K.

In [ ]:
protease_100Kdf = protease_df.query('`Collection Temperature` == 100')
protease_100Kdf

Let's check this new dataframe to make sure it does not contain any entries with data that were collected at temperatures other than 100K.

In [ ]:
protease_100Kdf.query('`Collection Temperature` != 100')

It looks like it worked - there are no entries in protease_100Kdf with a Collection Temperature that is not 100K.

**Challenge**

Create a new dataframe from protease_df that contains structures with resolutions less than 2.0 that had data collected at 298K


Now we have explored our dataframe with some analysis and queries based on numbers in of type float. How about strings? Can we find out how many of the structures in our dataframe contain the word "trypsin ". To see if a row in a dataframe contains as string, you can use the format

```python
df.series_label.str.contains('word')
```

In [29]:
# Query for whether lines in protease_df contain the word 'trypsin'

protease_df.Title.str.contains('trypsin')

0      False
1      False
2       True
3      False
4      False
       ...  
672    False
673     True
674    False
675    False
676    False
Name: Title, Length: 677, dtype: bool

All the lines that say `True` contain the word 'trypsin'. If we want to display the lines that contain the word 'trypsin' we can use the command below. The syntax here looks a bit complex, but it is simply telling the Jupyter notebook to display lines where this query is `True`.

```python
protease_df.Title.str.contains('trypsin')
```

In the subsequent cell, we can create a new dataframe that contains only structures that contain the word 'trypsin' in their titles.

In [30]:
# Query for all lines in protease_df that contain the word 'trypsin'

protease_df[protease_df.Title.str.contains('trypsin' or 'Trypsin')]

Entry ID  Collection Temperature Release Date  Resolution  \
2       2R9P                   100.0     12/11/07    1.400000   
9       2QYI                   100.0      8/19/08    2.600000   
10      2RA3                   100.0     12/11/07    1.460000   
27      6DI8                   100.0      1/23/19    1.859000   
32      6E5M                   100.0      3/13/19    1.612000   
38      6DZF                   100.0      9/19/18    2.200000   
39      6EAU                   100.0       8/7/19    1.180000   
40      6EAT                   100.0      3/13/19    1.149000   
41      6EAW                   100.0       8/7/19    1.289000   
42      6EAV                   100.0       8/7/19    1.391000   
43      6EAX                   100.0       8/7/19    1.189000   
81      6VXY                   100.0       7/1/20    1.398000   
100     2OTV                   292.0      2/27/07    1.560000   
102     2P8O                    77.2       5/8/07    1.500000   
108     2OXS                   300.0      3/13/07    1.320000   
170     6ZFK                   100.0      1/13/21    1.100000   
171     6ZFJ                   100.0      1/13/21    1.000000   
174     6ZQ2                   100.0      7/21/21    1.290000   
274     6XYG                   295.0      2/10/21    1.500002   
275     6XYK                   295.0      1/27/21    1.500000   
284     6YIW                   100.0      12/2/20    0.970000   
285     6YIV                   100.0      12/2/20    0.950000   
286     6YIY                   100.0      12/2/20    1.110000   
287     6YIX                   100.0      12/2/20    1.310000   
288     6YIS                   100.0      12/2/20    1.190000   
289     6YIU                   100.0      12/2/20    1.360000   
290     6YIT                   100.0      12/2/20    1.250000   
306     6YZA                   100.0      12/2/20    1.300000   
307     6YZC                   100.0      12/2/20    1.420000   
402     1P2J                   100.0      4/20/04    1.350000   
403     1P2I                   100.0      4/20/04    1.650000   
404     1P2Q                   100.0      4/20/04    1.800000   
405     1P2O                   100.0      4/20/04    2.000000   
406     1P2N                   100.0      4/20/04    1.800000   
407     1P2M                   100.0      4/20/04    1.750000   
408     1P2K                   100.0      4/20/04    1.600000   
417     1OX1                   291.0      5/18/04    2.000000   
427     1OXG                   278.0      5/18/04    2.200000   
566     2IJO                   170.0      5/15/07    2.300000   
568     2ILN                   100.0      4/10/07    2.000000   
628     6SV6                   100.0      1/22/20    1.150000   
629     6SV8                   100.0      1/22/20    1.150000   
630     6SV0                   100.0      1/22/20    1.160000   
632     6SVR                   100.0      1/22/20    1.160000   
633     6SVV                   100.0      1/22/20    1.150000   
634     6SVU                   100.0      1/22/20    1.150000   
635     6SVX                   100.0      1/22/20    1.160000   
637     6SVW                   100.0      1/22/20    1.160000   
638     6SVJ                   100.0      1/22/20    1.160000   
639     6SVI                   100.0      1/22/20    1.160000   
640     6SVN                   100.0      1/22/20    1.160000   
642     6SW0                   100.0      1/22/20    1.150000   
643     6SVB                   100.0      1/22/20    1.150000   
644     6SVD                   100.0      1/22/20    1.150000   
646     6SVG                   100.0      1/22/20    1.160000   
647     6SUX                   100.0      1/22/20    1.160000   
648     6SV9                   100.0      1/22/20    1.150000   
653     6SVZ                   100.0      1/22/20    1.150000   
673     7JRX                   100.0      7/21/21    1.770000   

                                                 Title  
2    Human mesotrypsin complexed with bovine p

In [31]:
# Creating a new dataframe that includes all rows from protease_df
# containing the word 'trypsin'

trypsin_df = protease_df[protease_df.Title.str.contains('trypsin' or 'Trypsin')]
trypsin_df

Entry ID  Collection Temperature Release Date  Resolution  \
2       2R9P                   100.0     12/11/07    1.400000   
9       2QYI                   100.0      8/19/08    2.600000   
10      2RA3                   100.0     12/11/07    1.460000   
27      6DI8                   100.0      1/23/19    1.859000   
32      6E5M                   100.0      3/13/19    1.612000   
38      6DZF                   100.0      9/19/18    2.200000   
39      6EAU                   100.0       8/7/19    1.180000   
40      6EAT                   100.0      3/13/19    1.149000   
41      6EAW                   100.0       8/7/19    1.289000   
42      6EAV                   100.0       8/7/19    1.391000   
43      6EAX                   100.0       8/7/19    1.189000   
81      6VXY                   100.0       7/1/20    1.398000   
100     2OTV                   292.0      2/27/07    1.560000   
102     2P8O                    77.2       5/8/07    1.500000   
108     2OXS                   300.0      3/13/07    1.320000   
170     6ZFK                   100.0      1/13/21    1.100000   
171     6ZFJ                   100.0      1/13/21    1.000000   
174     6ZQ2                   100.0      7/21/21    1.290000   
274     6XYG                   295.0      2/10/21    1.500002   
275     6XYK                   295.0      1/27/21    1.500000   
284     6YIW                   100.0      12/2/20    0.970000   
285     6YIV                   100.0      12/2/20    0.950000   
286     6YIY                   100.0      12/2/20    1.110000   
287     6YIX                   100.0      12/2/20    1.310000   
288     6YIS                   100.0      12/2/20    1.190000   
289     6YIU                   100.0      12/2/20    1.360000   
290     6YIT                   100.0      12/2/20    1.250000   
306     6YZA                   100.0      12/2/20    1.300000   
307     6YZC                   100.0      12/2/20    1.420000   
402     1P2J                   100.0      4/20/04    1.350000   
403     1P2I                   100.0      4/20/04    1.650000   
404     1P2Q                   100.0      4/20/04    1.800000   
405     1P2O                   100.0      4/20/04    2.000000   
406     1P2N                   100.0      4/20/04    1.800000   
407     1P2M                   100.0      4/20/04    1.750000   
408     1P2K                   100.0      4/20/04    1.600000   
417     1OX1                   291.0      5/18/04    2.000000   
427     1OXG                   278.0      5/18/04    2.200000   
566     2IJO                   170.0      5/15/07    2.300000   
568     2ILN                   100.0      4/10/07    2.000000   
628     6SV6                   100.0      1/22/20    1.150000   
629     6SV8                   100.0      1/22/20    1.150000   
630     6SV0                   100.0      1/22/20    1.160000   
632     6SVR                   100.0      1/22/20    1.160000   
633     6SVV                   100.0      1/22/20    1.150000   
634     6SVU                   100.0      1/22/20    1.150000   
635     6SVX                   100.0      1/22/20    1.160000   
637     6SVW                   100.0      1/22/20    1.160000   
638     6SVJ                   100.0      1/22/20    1.160000   
639     6SVI                   100.0      1/22/20    1.160000   
640     6SVN                   100.0      1/22/20    1.160000   
642     6SW0                   100.0      1/22/20    1.150000   
643     6SVB                   100.0      1/22/20    1.150000   
644     6SVD                   100.0      1/22/20    1.150000   
646     6SVG                   100.0      1/22/20    1.160000   
647     6SUX                   100.0      1/22/20    1.160000   
648     6SV9                   100.0      1/22/20    1.150000   
653     6SVZ                   100.0      1/22/20    1.150000   
673     7JRX                   100.0      7/21/21    1.770000   

                                                 Title  
2    Human mesotrypsin complexed with bovine p

<div class="exercise-this-is-a-title exercise">
<p class="exercise-title">Exercise</p>
    <p> The enzyme kinetics data from the introduction can be found in the file `enzyme_kinetics.csv` in the data directory. The only problem is that the substrate concentrations are out of order. This is how it looks in Excel:</p>
    
![Enzyme Kinetics](images/enzyme_kinetics.png)
        
Create a dataframe from that file and use pandas functions to place the data in ascending substrate concentrations, then print the velocity at a substrate concentration of 20 micromolar. As a biochemist, you probably want to plot and analyze these data to find the Km and Vmax. We'll get to that in a bit. The goal right now is to just have a little fun with pandas.
    
```{admonition} Solution
:class: dropdown
    import os
    import pandas as pd
    kinetics_file = os.path.join('data', 'enzyme_kinetics.csv')
    kinetics_df = pd.read_csv(kinetics_file) # kinetics_file is a variable, not a string
    kinetics_df.head(0)  # to find out the column labels. You can also use kinetics_df.columns to get the same output.
    kinetics_df.sort_values('Substrate_conc', inplace = True)  # once you know the column labels
    kinetics_df.loc[9]
```
    
</div>